blog/2023-09-25-constructs-methods.md

```diff
@@ -0,0 +1,247 @@
+const rose = new Flower(garden, "rose", {
+  kind: "rose",
+  color: "red",
+});
+new OrangePatch(garden, "orange-patch", { flower: rose });
+new PurplePatch(garden, "purple-patch", { flower: rose });
+```
+
+Since they both set the color of `rose`, one of them is going to override the decision of the other (in this case, the final rose will be purple).
+Uh oh!
+
+To avoid these kinds of issues, we usually follow these two rules when designing the methods of constructs.
+
+### Rule 1: Only add state, never subtract or update
+
+Methods should add state, not update or subtract state. If you're always adding state, then state that was configured or added earlier in the application won't get removed or overridden. The additions should also be [commutative](https://en.wikipedia.org/wiki/Commutative_property) - which means re-ordering them should not change the application's functional behavior.
+
+We can see an example of this rule with the `addEnvironment` method on `cloud.Function`:
+
+```js
+let fn = new cloud.Function(/* props */);
+fn.addEnvironment("DB_NAME", "orders");
+```
+
+If you try calling `addEnvironment` with the same string twice, it throws an error.
+Since environment variables can only be added, you can pass around `fn` throughout your codebase - including to third party libraries! - without worrying about environment variables being removed or changed.
+
+Besides changing internal state on a construct, methods can also create new auxiliary constructs, which is another way to safely add state. For example, in Wing it's possible to add objects to a bucket during deployment like so:
+
+```js
+bring cloud;
+
+let bucket = new cloud.Bucket();
+bucket.addObject("hello.txt", "cool beans");
+```
+
+Under the hood, `addObject` creates a new construct representing a bucket object, and that gets translated into IaC configuration separately from the actual bucket.
+
+### Rule 2: Document destructive APIs
+
+While methods that destroy existing state are worth avoiding, if there's a need for them, document the APIs accordingly.
+
+For example, if changing a flower's color is truly necessary, it's a good practice to give the method a descriptive name like `overrideColor()` to make it clear when reading the code that something exceptional is happening.
+
+Another common use case for mutating APIs are to provide [escape hatches](https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html#cfn_layer_raw) for when an abstraction doesn't expose all of the capabilities you need.
+You shouldn't need them often, but when you do, you're usually glad they're available.
+
+## To be continued
+
+By following the rules above, you'll find it easier to design more safer APIs like the classes in the AWS CDK and Wing's standard library that lead to fewer surprises when they're used by other developers.
+
+In the next post, we'll explore more design patterns in the constructs ecosystem, including how to safely design "getters" on constructs in the face of mutations.
```

Comment:
In my experience it is a bit tricky to promise a multi part series without having written it all already. It's a recipe for stress and frustration.

I am wondering if perhaps maybe the title of this post is "Avoiding mutability pitfalls in SDK design" and make this a bit more personal and "small": "I work on the Wing SDK and as part of this work, we've been working to design the APIs for the Wing Cloud Library. Here's one of the lessons I've learned when dealing with mutability"...

Then, if you end up writing more stuff around API design (which would be AWESOME!), then you can edit this post and chain them together into a series.

Reply:
Reworded the conclusion to avoid making any promises :)